beacon-chain/blockchain/process_block_helpers.go

```diff
@@ -129,6 +128,94 @@ func (s *Service) sendLightClientFeeds(cfg *postBlockProcessConfig) {
 	}
 }
 
+// saveLightClientUpdate saves the light client update for this block
+// if it's better than the already saved one, when feature flag is enabled.
+func (s *Service) saveLightClientUpdate(cfg *postBlockProcessConfig) {
+	//if !features.Get().EnableLightClient {
+	//	return
+	//}
+
+	attestedRoot := cfg.roblock.Block().ParentRoot()
+	attestedBlock, err := s.getBlock(cfg.ctx, attestedRoot)
+	if err != nil {
+		log.WithError(err).Error("Could not get attested block")
+		return
+	}
+	if attestedBlock == nil {
+		log.Error("Attested block is nil")
+		return
+	}
+	attestedState, err := s.cfg.StateGen.StateByRoot(cfg.ctx, attestedRoot)
+	if err != nil || attestedState == nil {
+		log.WithError(err).Error("Could not get attested state")
+		return
+	}
+
+	finalizedRoot := attestedState.FinalizedCheckpoint().Root
+	finalizedBlock, err := s.getBlock(cfg.ctx, [32]byte(finalizedRoot))
+	if err != nil {
+		log.WithError(err).Error("Could not get finalized block")
+		return
+	}
+
+	update, err := lightclient.NewLightClientUpdateFromBeaconState(
+		cfg.ctx,
+		s.CurrentSlot(),
+		cfg.postState,
+		cfg.roblock,
+		attestedState,
+		attestedBlock,
+		finalizedBlock,
+	)
+	if err != nil {
+		log.WithError(err).Error("Could not create light client update")
+		return
+	}
+
+	period := uint64(attestedState.Slot()) / (uint64(params.BeaconConfig().SlotsPerEpoch) * uint64(params.BeaconConfig().EpochsPerSyncCommitteePeriod))
+
+	oldUpdate, err := s.cfg.BeaconDB.LightClientUpdate(cfg.ctx, period)
+	if err != nil {
+		log.WithError(err).Error("Could not get current light client update")
+		return
+	}
+	if oldUpdate == nil {
+		if err := s.cfg.BeaconDB.SaveLightClientUpdate(cfg.ctx, period, update); err != nil {
+			log.WithError(err).Error("Could not save light client update")
+		}
+	} else {
+		isNewUpdateBetter, err := lightclient.IsBetterUpdate(update, oldUpdate)
+		if err != nil {
+			log.WithError(err).Error("Could not compare light client updates")
+			return
+		}
+		if isNewUpdateBetter {
+			if err := s.cfg.BeaconDB.SaveLightClientUpdate(cfg.ctx, period, update); err != nil {
+				log.WithError(err).Error("Could not save light client update")
+			}
+		}
+	}
```

Comment:
- I think there is a way to rearrange this code to deduplicate the call to `SaveLightClientUpdate`.
- It would be good to have a debug log when the old update is better. That may help in future debugging of why updates are not being saved properly to the DB.

Reply:
added the logs. but I could not rearrange the code in a nice way to deduplice the call to `SaveLightClientUpdate`. feel free to suggest changes.